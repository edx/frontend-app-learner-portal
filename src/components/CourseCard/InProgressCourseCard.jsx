import React from 'react';
import PropTypes from 'prop-types';

import BaseCourseCard from './BaseCourseCard';

const InProgressCourseCard = (props) => {
  const renderButtonLink = () => (
    <a className="btn btn-outline-primary" href={props.linkToCourse}>Continue Learning</a>
  );

  return (
    <BaseCourseCard buttonLink={renderButtonLink()} {...props} />
  );
};

InProgressCourseCard.propTypes = {
  linkToCourse: PropTypes.string.isRequired,
};

export default InProgressCourseCard;
