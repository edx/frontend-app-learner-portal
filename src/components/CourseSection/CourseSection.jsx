import React from 'react';
import PropTypes from 'prop-types';

const CourseSection = (props) => {
  const { component: Component, enrollments, title } = props;

  if (enrollments.length > 0) {
    return (
      <div className="course-section">
        <h2 className="mb-4">{title}</h2>
        {enrollments.map((courseData) => {
          const { status } = courseData;
          const defaultCardProps = {
            title: courseData.display_name,
            microMastersTitle: courseData.micromasters_title,
          };
          const cardProps = {};
          switch (status) {
            case 'in-progress':
              cardProps.endDate = courseData.end_date;
              cardProps.linkToCourse = courseData.resume_course_run_url;
              break;
            case 'upcoming':
              cardProps.startDate = courseData.start_date;
              cardProps.linkToCourse = courseData.course_run_url;
              break;
            case 'completed':
              cardProps.endDate = courseData.end_date;
              cardProps.linkToCourse = courseData.course_run_url;
              break;
            default:
              break;
          }
          return <Component {...defaultCardProps} {...cardProps} />;
        })}
      </div>
    );
  }

  return null;
};

CourseSection.propTypes = {
  component: PropTypes.element.isRequired,
  enrollments: PropTypes.arrayOf(PropTypes.shape({
    display_name: PropTypes.string.isRequired,
    micromasters_title: PropTypes.string,
    start_date: PropTypes.string,
    end_date: PropTypes.string,
    resume_course_run_url: PropTypes.string,
    course_run_url: PropTypes.string,
  })).isRequired,
  title: PropTypes.string.isRequired,
};

export default CourseSection;
