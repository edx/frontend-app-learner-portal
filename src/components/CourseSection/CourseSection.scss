@import "~@edx/edx-bootstrap/scss/edx/theme";

.collapsible {
  border: 0;

  &.open {
    border: 0;
    
    .collapsible-body {
      padding: 0;
      overflow: initial;
    }
  }

  .collapsible-title {
    @extend .h2;
  }

  .btn-collapsible {
    background: $white;
    border-radius: 0;
    border: 0;
    @extend .mb-2;

    &.open {
      border: 0;
    }

    &:hover {
      background: $white;
    }

    &:focus {
      box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25);
    }
  }

  .btn-collapsible,
  .collapsible-body {
    padding: 0;
  }
}
