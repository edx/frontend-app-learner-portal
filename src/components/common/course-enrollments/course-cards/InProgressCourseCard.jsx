import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { sendTrackEvent } from '@edx/frontend-analytics';

import BaseCourseCard from './BaseCourseCard';
import { MarkCompleteModal } from './mark-complete-modal';
import Notification from './Notification';

import { LayoutContext } from '../../layout';
import { camelCaseObject } from '../../../../common/utils';

const InProgressCourseCard = ({
  linkToCourse,
  courseRunId,
  title,
  notifications,
  ...rest
}) => {
  const { pageContext: { pageType } } = useContext(LayoutContext);
  const [isMarkCompleteModalOpen, setIsMarkCompleteModalOpen] = useState(false);

  const renderButtons = () => (
    <a
      className="btn btn-outline-primary btn-xs-block"
      href={linkToCourse}
      onClick={() => {
        sendTrackEvent('edx.learner_portal.course.continued', {
          course_run_id: courseRunId,
        });
      }}
    >
      Continue Learning
      <span className="sr-only">for {title}</span>
    </a>
  );

  const filteredNotifications = notifications.filter((notification) => {
    const now = moment();
    if (moment(notification.date).isBetween(now, moment(now).add('1', 'w'))) {
      return notification;
    }
    return false;
  });

  const getDropdownMenuItems = () => {
    // TODO: We should try to avoid this sort of conditional logic based on
    // `pageType`. Instead, ideally, we'd be able to pull the dropdown
    // menu items from a parent component's context or pass them down the
    // component tree.
    if (pageType !== 'pages.EnterprisePage') {
      return [];
    }
    return [{
      key: 'mark-complete',
      type: 'button',
      onClick: () => {
        setIsMarkCompleteModalOpen(true);
        sendTrackEvent('edx.learner_portal.course.mark_complete.modal.opened', {
          course_run_id: courseRunId,
        });
      },
      children: (
        <>
          Mark as complete
          <span className="sr-only">for {title}</span>
        </>
      ),
    }];
  };

  const handleMarkCompleteModalOnClose = () => {
    setIsMarkCompleteModalOpen(false);
    sendTrackEvent('edx.learner_portal.course.mark_complete.modal.closed', {
      course_run_id: courseRunId,
    });
  };

  const handleMarkCompleteModalOnSuccess = (response) => {
    const transformedResponse = camelCaseObject(response);
    sendTrackEvent('edx.learner_portal.course.mark_complete.saved', {
      course_run_id: courseRunId,
    });
    // eslint-disable-next-line no-console
    console.log(transformedResponse);
    // TODO: use the response here
  };

  const renderNotifications = () => {
    if (!filteredNotifications.length) {
      return null;
    }
    return (
      <div className="notifications">
        <ul
          className="list-unstyled mb-0"
          aria-label="course due dates"
          role="alert"
        >
          {filteredNotifications.map(notificationProps => (
            <Notification
              key={`notification-${notificationProps.url}-${notificationProps.date}`}
              courseRunId={courseRunId}
              {...notificationProps}
            />
          ))}
        </ul>
      </div>
    );
  };

  return (
    <BaseCourseCard
      type="in_progress"
      buttons={renderButtons()}
      dropdownMenuItems={getDropdownMenuItems()}
      title={title}
      linkToCourse={linkToCourse}
      courseRunId={courseRunId}
      {...rest}
    >
      {renderNotifications()}
      <MarkCompleteModal
        isOpen={isMarkCompleteModalOpen}
        courseTitle={title}
        courseLink={linkToCourse}
        courseId={courseRunId}
        onClose={handleMarkCompleteModalOnClose}
        onSuccess={handleMarkCompleteModalOnSuccess}
      />
    </BaseCourseCard>
  );
};

InProgressCourseCard.propTypes = {
  linkToCourse: PropTypes.string.isRequired,
  courseRunId: PropTypes.string.isRequired,
  notifications: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
  })).isRequired,
  title: PropTypes.string.isRequired,
};

export default InProgressCourseCard;
