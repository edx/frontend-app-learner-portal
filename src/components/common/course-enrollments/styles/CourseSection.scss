@import "~@edx/paragon/scss/edx/theme.scss";

.collapsible-trigger,
.collapsible-body {
  @extend .px-0;
}
