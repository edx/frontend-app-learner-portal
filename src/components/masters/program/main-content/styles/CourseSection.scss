@import "~@edx/paragon/scss/edx/theme.scss";

.collapsible {
  @extend .border-0;
  
  .collapsible-body {
    @extend .p-0;

    &.open {
      overflow: initial;
    }
  }

  .collapsible-title {
    @extend .h2;
  }

  .btn-collapsible {
    @extend .rounded-0;
    @extend .mb-2;
    @extend .border-0;

    background-color: $white;

    &:hover {
      background-color: $white !important;
    }

    &:focus {
      box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25);
    }
  }

  .btn-collapsible,
  .collapsible-body {
    padding: 0;
  }
}
