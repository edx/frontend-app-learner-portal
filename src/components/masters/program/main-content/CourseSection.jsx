import React from 'react';
import PropTypes from 'prop-types';
import { sendTrackEvent } from '@edx/frontend-analytics';
import { Collapsible } from '@edx/paragon';
import { faChevronCircleUp, faChevronCircleDown } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { InProgressCourseCard, UpcomingCourseCard, CompletedCourseCard } from './course-cards';

import './styles/CourseSection.scss';

class CourseSection extends React.Component {
  state = {
    isOpen: true,
  };

  getFormattedTitle = () => {
    const { isOpen } = this.state;
    const { enrollments, title } = this.props;
    return isOpen ? title : `${title} (${enrollments.length})`;
  };

  handleCollapsibleToggle = (isOpen) => {
    this.setState({
      isOpen,
    });
    sendTrackEvent('edx.learner_portal.section.toggled', { is_open: isOpen, section_title: this.props.title });
  };

  render() {
    const { component: Component, enrollments } = this.props;

    if (enrollments && enrollments.length > 0) {
      return (
        <div className="course-section mb-5">
          <Collapsible
            title={this.getFormattedTitle()}
            onToggle={this.handleCollapsibleToggle}
            icons={{
              expanded: <FontAwesomeIcon className="text-primary" icon={faChevronCircleUp} />,
              collapsed: <FontAwesomeIcon className="text-primary" icon={faChevronCircleDown} />,
            }}
            isOpen
          >
            {enrollments.map((courseData) => {
              const { course_run_status: courseRunStatus } = courseData;
              const defaultCardProps = {
                key: courseData.course_run_id,
                courseRunId: courseData.course_run_id,
                title: courseData.display_name,
                microMastersTitle: courseData.micromasters_title,
                hasEmailsEnabled: courseData.emails_enabled,
              };
              const cardProps = {};
              switch (courseRunStatus) {
                case 'in_progress':
                  cardProps.endDate = courseData.end_date;
                  // The link to course here gives precedence to the resume course link, which is
                  // present if the learner has made progress. If the learner has not made progress,
                  // we should link to the main course run URL.
                  cardProps.linkToCourse = courseData.resume_course_run_url ||
                    courseData.course_run_url;
                  cardProps.notifications = courseData.due_dates;
                  break;
                case 'upcoming':
                  cardProps.startDate = courseData.start_date;
                  cardProps.linkToCourse = courseData.course_run_url;
                  break;
                case 'completed':
                  cardProps.endDate = courseData.end_date;
                  cardProps.linkToCourse = courseData.course_run_url;
                  cardProps.linkToCertificate = courseData.certificate_download_url;
                  break;
                default:
                  break;
              }
              return <Component {...defaultCardProps} {...cardProps} />;
            })}
          </Collapsible>
        </div>
      );
    }

    return null;
  }
}

CourseSection.propTypes = {
  component: PropTypes.oneOf([
    InProgressCourseCard,
    UpcomingCourseCard,
    CompletedCourseCard,
  ]).isRequired,
  enrollments: PropTypes.arrayOf(PropTypes.shape({
    display_name: PropTypes.string.isRequired,
    micromasters_title: PropTypes.string,
    start_date: PropTypes.string,
    end_date: PropTypes.string,
    resume_course_run_url: PropTypes.string,
    course_run_url: PropTypes.string,
  })).isRequired,
  title: PropTypes.string.isRequired,
};

export default CourseSection;
