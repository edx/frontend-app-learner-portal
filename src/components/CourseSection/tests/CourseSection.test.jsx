import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import CourseSection from '../CourseSection';
import InProgressCourseCard from '../../CourseCard/InProgressCourseCard';

const sampleEnrollmentData = {
  course_run_id: 'edX+DemoX+Demo_Course',
  display_name: 'edX Demonstration Course',
  resume_course_run_url: 'https://edx.org/',
  course_run_url: 'https://edx.org/',
  start_date: '2017-02-05T05:00:00Z',
  end_date: '2019-08-18T05:00:00Z',
  emails_enabled: true,
  due_dates: [],
  micromasters_title: null,
  certificate_generation_url: null,
  status: 'in-progress',
};

describe('<CourseSection />', () => {
  it('does not render section if there are no enrollments', () => {
    const tree = renderer
      .create((
        <CourseSection
          title="Example Title"
          component={InProgressCourseCard}
          enrollments={[]}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders section if there are enrollments', () => {
    const tree = renderer
      .create((
        <CourseSection
          title="Example Title"
          component={InProgressCourseCard}
          enrollments={[sampleEnrollmentData, {
            ...sampleEnrollmentData,
            course_run_id: 'edX+DemoX+Demo_Course_2',
            display_name: 'edX Demonstration Course 2',
          }]}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('properly handles collapsible behavior', () => {
    const sectionTitle = 'Example Title';
    const enrollments = [sampleEnrollmentData];
    const wrapper = mount((
      <CourseSection
        title={sectionTitle}
        component={InProgressCourseCard}
        enrollments={enrollments}
      />
    ));
    expect(wrapper.state('isOpen')).toBeTruthy();
    expect(wrapper.find('.collapsible-title').text()).toEqual(sectionTitle);

    wrapper.find('.btn-collapsible').first().simulate('click');

    expect(wrapper.state('isOpen')).toBeFalsy();
    expect(wrapper.find('.collapsible-title').text()).toEqual(`${sectionTitle} (${enrollments.length})`);
  });
});
