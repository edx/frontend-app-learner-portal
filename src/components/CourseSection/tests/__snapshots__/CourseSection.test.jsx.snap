// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CourseSection /> does not render section if there are no enrollments 1`] = `null`;

exports[`<CourseSection /> renders section if there are enrollments 1`] = `
<div
  className="course-section mb-5"
>
  <div
    className="collapsible open"
  >
    <button
      aria-expanded={true}
      className="btn btn-block text-left btn-collapsible open"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      <div
        className="collapsible-title d-flex align-items-center justify-content-between"
      >
        Example Title
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-circle-up fa-w-16 text-primary"
          data-icon="chevron-circle-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 256C8 119 119 8 256 8s248 111 248 248-111 248-248 248S8 393 8 256zm231-113.9L103.5 277.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L256 226.9l101.6 101.6c9.4 9.4 24.6 9.4 33.9 0l17-17c9.4-9.4 9.4-24.6 0-33.9L273 142.1c-9.4-9.4-24.6-9.4-34 0z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </button>
    <div
      className="collapsible-body open"
    >
      <div
        className="card mb-4"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters mb-3"
          >
            <div
              className="col-lg-12 col-xl-8"
            >
              <h3
                className="card-title mb-1 font-weight-normal"
              >
                <a
                  href="https://edx.org/"
                >
                  edX Demonstration Course
                </a>
              </h3>
              <p
                className="card-text"
              >
                Course 
                ends
                 on 
                August 18, 2019
              </p>
            </div>
            <div
              className="col-lg-12 col-xl-4 text-xl-right mt-3 mt-xl-0"
            >
              <a
                className="btn btn-primary btn-block"
                href="https://edx.org/"
              >
                Continue Learning
              </a>
            </div>
          </div>
          <div
            className="row no-gutters"
          >
            <div
              className="col"
            >
              <button
                className="email-settings-btn btn btn-link p-0"
                onClick={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-cog fa-w-16 mr-2"
                  data-icon="cog"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                Email settings
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card mb-4"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters mb-3"
          >
            <div
              className="col-lg-12 col-xl-8"
            >
              <h3
                className="card-title mb-1 font-weight-normal"
              >
                <a
                  href="https://edx.org/"
                >
                  edX Demonstration Course 2
                </a>
              </h3>
              <p
                className="card-text"
              >
                Course 
                ends
                 on 
                August 18, 2019
              </p>
            </div>
            <div
              className="col-lg-12 col-xl-4 text-xl-right mt-3 mt-xl-0"
            >
              <a
                className="btn btn-primary btn-block"
                href="https://edx.org/"
              >
                Continue Learning
              </a>
            </div>
          </div>
          <div
            className="row no-gutters"
          >
            <div
              className="col"
            >
              <button
                className="email-settings-btn btn btn-link p-0"
                onClick={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-cog fa-w-16 mr-2"
                  data-icon="cog"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                Email settings
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
